File: ch5/src/py_demo_tf/py_demo_tf/tf_listener.py
import rclpy
from rclpy.node import Node
from tf2_ros import TransformListener, Buffer
from tf_transformations import euler_from_quaternion

class TFListener(Node):
    def __init__(self):
        super().__init__('tf_listener')
        self.buffer_ = Buffer()
        self.listener_ = TransformListener(self.buffer_, self)
        self.timers_ = self.create_timer(1, self.get_transform)
        
    def get_transform(self):
        try:
            result = self.buffer_.lookup_transform(
                'base_link',
                'bottle_link',
                rclpy.time.Time(seconds=0), 
                rclpy.time.Duration(seconds=1)
            )
            
            transform = result.transform
            rotation_euler = euler_from_quaternion(
                [
                    transform.rotation.x,
                    transform.rotation.y,
                    transform.rotation.z,
                    transform.rotation.w
                ]
            )    
            self.get_logger().info(f"translation: {transform.translation}, quaternion: {transform.rotation}, euler: {rotation_euler}")
        except Exception as e:
            self.get_logger().warn(f"can not get transform : {e}")
            

def main():
    rclpy.init()
    tf_listener = TFListener()
    rclpy.spin(tf_listener)
    rclpy.shutdown()

File: ch5/src/cpp_demo_tf/src/tf_listener.cpp
#include <memory>
#include <chrono>
#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/transform_stamped.hpp"
#include "tf2/LinearMath/Quaternion.h"
#include "tf2_geometry_msgs/tf2_geometry_msgs.hpp"
#include "tf2_ros/transform_listener.h"
#include "tf2_ros/buffer.h"
#include "tf2/utils.h"

using namespace std::chrono_literals;

class TFListener: public rclcpp::Node {
private:
    std::shared_ptr<tf2_ros::Buffer> buffer_;
    std::shared_ptr<tf2_ros::TransformListener> listener_;
    rclcpp::TimerBase::SharedPtr timer_;

public:
    TFListener(): Node("tf_listener") {
        this->buffer_ = std::make_shared<tf2_ros::Buffer>(this->get_clock());
        this->listener_ = std::make_shared<tf2_ros::TransformListener>(*buffer_, this);
        this->timer_ = this->create_wall_timer(5s, std::bind(&TFListener::get_transform, this));
    }
    void get_transform() {
        try {
            const auto transform = buffer_->lookupTransform(
                "base_link", "target_point",
                this->get_clock()->now(),
                rclcpp::Duration::from_seconds(1.0f)
            );

            const auto& translation = transform.transform.translation;
            const auto& rotation = transform.transform.rotation;
            double r, p, y;
            tf2::getEulerYPR(rotation, y, p, r);
            RCLCPP_INFO(this->get_logger(), "translation: (%f, %f, %f)", translation.x, translation.y, translation.z);
            RCLCPP_INFO(this->get_logger(), "rotation: (%f, %f, %f)", r, p, y);

        } catch (tf2::TransformException& ex) {
            RCLCPP_WARN(this->get_logger(), "exception: {%s}", ex.what());
        }
    }
};


int main(int argc, char** argv) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<TFListener>();
    rclcpp::spin(node);
    rclcpp::shutdown();

    return 0;
}